### PR TITLE
Support segmentation mask in cutmix preprocessing layer

### LINKED ISSUE #1549: Support segmentation masks in `CutMix` layer
This should follow the same structure as segmentation mask augmentation in our other preprocessing layers.

Treating masks like 1-channel images for the purpose of applying CutMix should be a simple and effective approach for this.

### PR DESCRIPTION
**Pull request by cosmo3769**
# What does this PR do?

Fixes #1549 

pytest output:

Screenshot 2023-07-07 at 8 23 36 AM

demo: 

Screenshot 2023-07-03 at 8 17 40 PM
 

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case. #1549 
- [ ] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

@ianstenbit @jbischof @LukeWood

### CODE REVIEW

File: keras_cv/layers/preprocessing/cut_mix.py [hunk 1]
```diff
@@ -31,24 +31,28 @@ class CutMix(BaseImageAugmentationLayer):
...
-    # output == {'images': updated_images, 'labels': updated_labels}
-    ```
     """
 
-    def __init__(self, alpha=1.0, seed=None, **kwargs):
+    def __init__(
+        self,
+        alpha=1.0,
+        seed=None,
+        apply_to_labels=False,
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/cut_mix.py:47]: Let's drop both of these parameters -- we should automatically apply cutmix to all inputs to the layer (including labels and masks)

File: keras_cv/layers/preprocessing/cut_mix_test.py [hunk 2]
```diff
@@ -79,6 +100,89 @@ def test_cut_mix_call_results_one_channel(self):
...
+        )
+
+        ys_segmentation_masks = tf.cast(
+            tf.stack(
+                [2 * tf.ones((4, 4)), tf.ones((4, 4))],
+                axis=0,
+            ),
+            tf.uint8,
+        )
+        ys_segmentation_masks = tf.one_hot(ys_segmentation_masks, 3)
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/cut_mix_test.py:119]: We should also make sure the layer can handle dense-encoded segmentation masks (e.g that have shape (height, width, 1) or (height, width)

File: keras_cv/layers/preprocessing/cut_mix_test.py [hunk 3]
```diff
@@ -79,6 +100,89 @@ def test_cut_mix_call_results_one_channel(self):
...
+        self.assertTrue(tf.math.reduce_any(xs[0] == 2.0))
+        self.assertTrue(tf.math.reduce_any(xs[1] == 1.0))
+        self.assertTrue(tf.math.reduce_any(xs[1] == 2.0))
+        # At least some pixels should be replaced in the segmentation_masks
+        self.assertTrue(tf.math.reduce_any(ys_segmentation_masks[0] == 1.0))
+        self.assertTrue(tf.math.reduce_any(ys_segmentation_masks[0] == 0.0))
+        self.assertTrue(tf.math.reduce_any(ys_segmentation_masks[1] == 1.0))
+        self.assertTrue(tf.math.reduce_any(ys_segmentation_masks[1] == 0.0))
+
+    def test_cut_mix_call_results_one_channel_with_masks(self):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/cut_mix_test.py:141]: This test case doesn't seem to add a lot of coverage, so we can probably drop it.

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks for the PR -- this is great

Just a few comments for you 😃

File: keras_cv/layers/preprocessing/cut_mix_test.py (hunk 2 above)
Comment (cosmo3769) [reply keras_cv/layers/preprocessing/cut_mix_test.py]: Now, with the changes, it can handle both dense encoded and one-hot encoded segmentation masks.

File: keras_cv/layers/preprocessing/cut_mix.py (hunk 1 above)
Comment (cosmo3769) [reply keras_cv/layers/preprocessing/cut_mix.py]: Removed the parameters.

**Comment by cosmo3769** [comment]
I will fix the conflicts as new code of adding support for PyTorch and JAX has been implemented in this.

**Comment by ianstenbit** [comment]
> I will fix the conflicts as new code of adding support for PyTorch and JAX has been implemented in this.

SG -- there are some merge conflicts, but keep in mind that right now, our preprocessing layers are TF-only (they are intended to run in the TF.data graph).

In the long term they'll support Jax+PyTorch but for now no need to do that for this PR

**Comment by cosmo3769** [comment]
Merge conflict resolved. Please check. Thank you.

File: examples/layers/preprocessing/segmentation/demo_utils.py [hunk 4]
```diff
@@ -44,7 +44,6 @@ def to_dict(datapoint):
     input_mask = tf.one_hot(
         tf.squeeze(tf.cast(input_mask, tf.int32), axis=-1), depth=3
     )
-    input_image = tf.transpose(input_image, (2, 0, 1))
```
Comment (ianstenbit) [inline examples/layers/preprocessing/segmentation/demo_utils.py:47]: What is the purpose of dropping this? Does it break other segmentation demos?

**Review by ianstenbit — APPROVED** [review]
LGTM -- just a few little finishing touches. Thank you!

File: examples/layers/preprocessing/segmentation/demo_utils.py (hunk 4 above)
Comment (cosmo3769) [reply examples/layers/preprocessing/segmentation/demo_utils.py:47]: When transposing an image, cutmix operation does not work. Also, I checked with other segmentation demos, it works perfectly too.

File: keras_cv/layers/preprocessing/cut_mix.py [hunk 5]
```diff
@@ -31,21 +31,21 @@ class CutMix(BaseImageAugmentationLayer):
             the smoothing values are sampled. Defaults to 1.0, which is a
             recommended value when training an imagenet1k classification model.
         seed: Integer. Used to create a random seed.
+        apply_to_labels: Boolean. Cutmix applies to labels with images.
```
[the remark this answers is not in the document]
Comment (cosmo3769) [reply keras_cv/layers/preprocessing/cut_mix.py]: Ah! yes.
